be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/validators/ElectionValidator.scala

```diff
@@ -113,26 +121,95 @@ sealed class ElectionValidator(dbActorRef: => AskableActorRef) extends MessageDa
         val electionId: Hash = channel.extractChildChannel
         val sender: PublicKey = message.sender
 
-        val laoId: Hash = channel.decodeChannelLaoId.getOrElse(HASH_ERROR)
+        val laoId: Hash = channel.decodeChannelLaoId getOrElse HASH_ERROR
+
+        val setupMessage: SetupElection = getSetupMessage(channel, dbActor)
+        val questions = setupMessage.questions
+        val q2Ballots = questions.map(question => question.id -> question.ballot_options).toMap
 
         if (!validateTimestampStaleness(data.created_at)) {
           Right(validationError(s"stale 'created_at' timestamp (${data.created_at})"))
         } else if (electionId != data.election) {
           Right(validationError("unexpected election id"))
-        } else if (laoId != data.lao){
+        } else if (laoId != data.lao) {
           Right(validationError("unexpected lao id"))
-        } else if (!validateAttendee(sender, channel, dbActorRef)) {
+          //  check it the question id exists
+        } else if (!data.votes.map(_.question).forall(question => q2Ballots.contains(question))) {
+          Right(validationError(s"Incorrect parameter questionId"))
+        } else if (!data.votes.forall(
+          voteElection => {
+            val vote = voteElection.vote match {
+              case Some(v :: _) => v
+              case _ => -1
+            }
+            // check if the ballot is available
+            vote < q2Ballots(voteElection.question).size
+          }
+        )) {
+          Right(validationError(s"Incorrect parameter ballot"))
+          // check for question id duplication
+        } else if (data.votes.map(_.question).distinct.length != data.votes.length) {
+          Right(validationError(s"The castvote contains twice the same question id"))
+          // check open and end constraints
+        } else if (getOpenMessage(channel).isEmpty) {
+          Right(validationError(s"This election has not started yet"))
+        } else if (getEndMessage(channel).isDefined) {
+          Right(validationError(s"This election has already ended"))
+        } else if (!validateAttendee(sender, channel, dbActor)) {
           Right(validationError(s"Sender $sender has an invalid PoP token."))
-        } else if (!validateChannelType(ObjectType.ELECTION, channel, dbActorRef)) {
+        } else if (!validateChannelType(ObjectType.ELECTION, channel, dbActor)) {
           Right(validationError(s"trying to send a CastVoteElection message on a wrong type of channel $channel"))
         } else {
           Left(rpcMessage)
         }
-
       case _ => Right(validationErrorNoMessage(rpcMessage.id))
     }
   }
 
+  private def getOpenMessage(electionChannel: Channel): Option[OpenElection] = {
+    var result: Option[OpenElection] = None
+    Await.ready(dbActor ? DbActor.ReadChannelData(electionChannel), duration).value match {
```

Comment:
Might have been mentioned at the last meeting but this should probably be a `flatMap`.